server/src/main/java/org/elasticsearch/common/logging/DeprecatedMessage.java

```diff
@@ -0,0 +1,33 @@
+/*
+ * Licensed to Elasticsearch under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.common.logging;
+
+import java.util.Map;
+
+/**
+ * A logger message used by {@link DeprecationLogger}.
+ * Carries x-opaque-id field if provided in the headers. Will populate the x-opaque-id field in JSON logs.
+ */
+public class DeprecatedMessage extends ESLogMessage {
+
+    public DeprecatedMessage(String messagePattern, String xOpaqueId, Object... args) {
+        super(Map.of("x-opaque-id", inQuotes(xOpaqueId)), messagePattern, args);
```

Comment:
Can we not have a deprecated message without an opaque id now? I would rather if we could only add it to the log if it exists as a header. Otherwise the deprecation log for a user that doesn't know about it is just filled with an extra key on every deprecation message that is empty.

Reply:
We can do this. The implementation will end up slightly more complex, but indeed there is no point for this field to be present when empty

Reply:
re https://github.com/elastic/elasticsearch/pull/41354#discussion_r298457459
> This test is still commented out. We should still make it work?
absolutely, thanks for reminding!